File: godtools-tract-renderer/src/test/java/org/cru/godtools/tract/model/ImageGravityTest.java
package org.cru.godtools.tract.model;

import org.junit.Test;

import static org.cru.godtools.tract.model.ImageGravity.BOTTOM;
import static org.cru.godtools.tract.model.ImageGravity.CENTER;
import static org.cru.godtools.tract.model.ImageGravity.END;
import static org.cru.godtools.tract.model.ImageGravity.START;
import static org.cru.godtools.tract.model.ImageGravity.TOP;
import static org.cru.godtools.tract.model.ImageGravity.isBottom;
import static org.cru.godtools.tract.model.ImageGravity.isCenter;
import static org.cru.godtools.tract.model.ImageGravity.isCenterX;
import static org.cru.godtools.tract.model.ImageGravity.isCenterY;
import static org.cru.godtools.tract.model.ImageGravity.isEnd;
import static org.cru.godtools.tract.model.ImageGravity.isStart;
import static org.cru.godtools.tract.model.ImageGravity.isTop;
import static org.cru.godtools.tract.model.ImageGravity.parse;
import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.CoreMatchers.nullValue;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertThat;
import static org.junit.Assert.assertTrue;

@SuppressWarnings("ConstantConditions")
public class ImageGravityTest {
    @Test
    public void verifyParse() throws Exception {
        int gravity = parse("start unrecognized center", null);
        assertFalse(isCenterX(gravity));
        assertTrue(isStart(gravity));
        assertFalse(isEnd(gravity));
        assertTrue(isCenterY(gravity));
        assertFalse(isTop(gravity));
        assertFalse(isBottom(gravity));
        assertFalse(isCenter(gravity));

        gravity = parse("center end", null);
        assertFalse(isCenterX(gravity));
        assertFalse(isStart(gravity));
        assertTrue(isEnd(gravity));
        assertTrue(isCenterY(gravity));
        assertFalse(isTop(gravity));
        assertFalse(isBottom(gravity));
        assertFalse(isCenter(gravity));

        gravity = parse("center", null);
        assertTrue(isCenterX(gravity));
        assertFalse(isStart(gravity));
        assertFalse(isEnd(gravity));
        assertTrue(isCenterY(gravity));
        assertFalse(isTop(gravity));
        assertFalse(isBottom(gravity));
        assertTrue(isCenter(gravity));
    }

    @Test
    public void verifyParseConflictingGravity() throws Exception {
        assertThat(parse("start end", null), nullValue());
        assertThat(parse("start top end", null), nullValue());
        assertThat(parse("bottom top end", null), nullValue());
    }

    @Test
    public void verifyParseDefaultValue() throws Exception {
        for (final int align : new int[] {CENTER, START, END, TOP, BOTTOM}) {
            assertThat(parse(null, align), is(align));
            assertThat(parse("invalid-type", align), is(align));
        }
    }
}

File: library/xml-model/src/test/java/org/cru/godtools/xml/model/ImageGravityTest.kt
package org.cru.godtools.xml.model

import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.equalTo
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class ImageGravityTest {
    private val none = ImageGravity(0)

    @Test
    fun verifyParse() {
        ImageGravity.parse("start unrecognized center", ImageGravity.NONE).also { gravity ->
            assertFalse(gravity.isCenterX)
            assertTrue(gravity.isStart)
            assertFalse(gravity.isEnd)
            assertTrue(gravity.isCenterY)
            assertFalse(gravity.isTop)
            assertFalse(gravity.isBottom)
            assertFalse(gravity.isCenter)
        }

        ImageGravity.parse("center end", ImageGravity.NONE).also { gravity ->
            assertFalse(gravity.isCenterX)
            assertFalse(gravity.isStart)
            assertTrue(gravity.isEnd)
            assertTrue(gravity.isCenterY)
            assertFalse(gravity.isTop)
            assertFalse(gravity.isBottom)
            assertFalse(gravity.isCenter)
        }

        ImageGravity.parse("center", ImageGravity.NONE).also { gravity ->
            assertTrue(gravity.isCenterX)
            assertFalse(gravity.isStart)
            assertFalse(gravity.isEnd)
            assertTrue(gravity.isCenterY)
            assertFalse(gravity.isTop)
            assertFalse(gravity.isBottom)
            assertTrue(gravity.isCenter)
        }
    }

    @Test
    fun verifyParseConflictingGravity() {
        assertEquals(none, ImageGravity.parse("start end", none))
        assertEquals(none, ImageGravity.parse("start top end", none))
        assertEquals(none, ImageGravity.parse("bottom top end", none))
    }

    @Test
    fun verifyParseDefaultValue() {
        arrayOf(
            ImageGravity.START,
            ImageGravity.END,
            ImageGravity.TOP,
            ImageGravity.BOTTOM,
            ImageGravity.CENTER
        ).forEach {
            assertThat(ImageGravity.parse(null, it), equalTo(it))
            assertThat(ImageGravity.parse("invalid-type", it), equalTo(it))
        }
    }
}
